### PR TITLE
[AY1920S1-CS2113-T13-4] Duke Manager

### PR DESCRIPTION
**Pull request by maxxyx96**
@talesrune
@e0318465
@dou-maokang
@gervaiseang
@maxxyx96

### CODE REVIEW

**Comment by gervaiseang** [comment]
Implemented B-Reminders v1.0 but still left minor edits.

File: src/main/java/duke/AddWindow.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+     *
+     * @param d The object of Duke.
+     * @param mainWindow The main window that runs DUKE Manager.
+     */
+    @FXML
+    public void setAddWindow(Duke d, MainWindow mainWindow) {
+        duke = d;
+        this.mainWindow = mainWindow;
+        cbType.getItems().addAll(
+                "Todo",
```
Comment (RubaP) [inline src/main/java/duke/AddWindow.java]: Define these magical strings as constants (variable or enum or whatever the effective way)

File: src/main/java/duke/AddWindow.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                "Do After"
+        );
+        TaskList items = d.getTaskList();
+        for (int i = ZERO; i < items.size(); i++) {
+            cbExistingTask.getItems().add(items.get(i).getDescription());
+        }
+    }
+
+    @FXML
+    void onMouseClickAdd() {
```
Comment (RubaP) [inline src/main/java/duke/AddWindow.java]: Again please handle all the magical strings in this method as constants

File: src/main/java/duke/AddWindow.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        System.out.println("clicked on " + cbType.getSelectionModel().getSelectedItem());
+
+        if (cbType.getSelectionModel().getSelectedItem().equals("Todo")) {
+            mainWindow.handleUserEvent("todo " + tfDesc.getText().trim());
```
Comment (RubaP) [inline src/main/java/duke/AddWindow.java]: Why does onMouseClickAdd method deal with string manipulations?

File: src/main/java/duke/AddWindow.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            mainWindow.handleUserEvent("doafter " + tfDesc.getText().trim()
+                    + " /after " +  cbExistingTask.getSelectionModel().getSelectedItem());
+        }
+        mainWindow.listViewRefresh();
+        Stage stage = (Stage) btnAdd.getScene().getWindow();
+        stage.close();
+    }
+
+    @FXML
+    void onMouseClickReset() {
```
Comment (RubaP) [inline src/main/java/duke/AddWindow.java]: A good example for following SLAP

File: src/main/java/duke/DialogBox.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        getChildren().setAll(tmp);
+        setAlignment(Pos.TOP_LEFT);
+    }
+
+    public static DialogBox getUserDialog(String text, Image img) {
+        return new DialogBox(text, img);
+    }
+
+    /**
+     * Hi.
```
Comment (RubaP) [inline src/main/java/duke/DialogBox.java]: Why is there a 'Hi.' here?

File: src/main/java/duke/Duke.java [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    /**
+     * Creates a duke to initialize storage, task list, and ui.
+     *
+     * @param filePath1 The location of the text file.
+     * @param filePath2 The location of the priority text file.
+     * @param filePathForBudget The location of the budget text file
+     * @param filePathForContacts The location of the contact text file.
+     */
+    public Duke(String filePath1, String filePath2, String filePathForBudget, String filePathForContacts) {
+        ui = new Ui();
```
Comment (RubaP) [inline src/main/java/duke/Duke.java]: This violates SLAP and the readability is very poor since all the important objects creations and error handling happens here. Regarding SLAP, this method does ui.methodCall() and very lower level operation of creating objects as well.

File: src/main/java/duke/Duke.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+            sentence = ui.readCommand();
+            ui.showLine(); //Please do not remove!
+            try {
+                Command cmd = Parser.parse(sentence, items, budgetList);
+                if (cmd instanceof ExitCommand) {
+                    priorityStorage.write(priorityList);
+                    budgetStorage.write(budgetList);
+                    cmd.executeStorage(items, ui, storage);
+                    break;
+                } else if (cmd instanceof ListPriorityCommand
```
Comment (RubaP) [inline src/main/java/duke/Duke.java]: Should the run method of Duke which is your starting point understand different low-level commands and react to it. I guess the responsibility fo run is to initiate the run flow by initiating the relevant objects. SRP is violated here

File: src/main/java/duke/HelpWindow.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+     *
+     * @param d The object of Duke.
+     * @param mainWindow The main window that runs DUKE Manager.
+     */
+    @FXML
+    public void setHelpWindow(Duke d, MainWindow mainWindow) {
+        duke = d;
+        this.mainWindow = mainWindow;
+        cbCommands.getItems().addAll(
+                "todo",
```
Comment (RubaP) [inline src/main/java/duke/HelpWindow.java]: Again the magical strings

File: src/main/java/duke/HelpWindow.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+                "doafter"
+        );
+
+        /**
+         * Create action event when command is selected from drop down list
+         */
+        EventHandler<ActionEvent> event = new EventHandler<ActionEvent>() {
+            @Override
+            public void handle(ActionEvent actionEvent) {
+                if (cbCommands.getSelectionModel().getSelectedItem().equals("todo")) {
```
Comment (RubaP) [inline src/main/java/duke/HelpWindow.java]: How about a switch case which will be more clear for the reader?

File: src/main/java/duke/MainWindow.java [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+    private TextField tfnewDateTime;
+    @FXML
+    private ComboBox<String> cbtaskType;
+
+    private Duke duke;
+
+    private static final int ZERO = 0;
+    private static final int ONE = 1;
+    private static final int TIMER_DELAY = 500;
+    private static final int VBOX_WIDTH = 200;
```
Comment (RubaP) [inline src/main/java/duke/MainWindow.java]: Good

File: src/main/java/duke/MainWindow.java [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+                    DialogBox.getDukeDialog(response, dukeImage)
+            );
+            e.printStackTrace();
+        }
+        userInput.clear();
+    }
+
+    @FXML
+    protected void handleUserEvent(String input) {
+        //String input = userInput.getText();
```
Comment (RubaP) [inline src/main/java/duke/MainWindow.java]: Please remove commented code

File: src/main/java/duke/MainWindow.java [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+                    DialogBox.getDukeDialog(response, dukeImage)
+            );
+            e.printStackTrace();
+        }
+        userInput.clear();
+    }
+
+    @FXML
+    private void onMouseClick_ListView() {
+        //System.out.println("clicked on " + listT.getSelectionModel().getSelectedItem());
```
Comment (RubaP) [inline src/main/java/duke/MainWindow.java]: Again, please remove all the commented code

File: src/main/java/duke/MainWindow.java [hunk 13]
```diff
@@ -0,0 +1,392 @@
...
+                DialogBox.getUserDialog(input, userImage)
+        );
+        try {
+            Command cmd = duke.getCommand(input);
+            duke.saveState(cmd);
+            response = Ui.showByeGui();
+            dialogContainer.getChildren().add(
+                    DialogBox.getDukeDialog(response, dukeImage)
+            );
+            timer.schedule(exitDuke, new Date(System.currentTimeMillis() + TIMER_DELAY));
```
Comment (RubaP) [inline src/main/java/duke/MainWindow.java]: Again you are violating SLAP. If the responsibility of this method is to call relevant object's functionality to exist the program then it should not do lower-level string manipulation tasks to follow SLAP

File: src/main/java/duke/command/AddBudgetCommand.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+     * Executes the command to add a certain amount to the existing budget.
+     *
+     * @param items The task list that contains a list of tasks.
+     * @param ui    To tell the user that it is executed successfully.
+     */
+    @Override
+    public void execute(TaskList items, Ui ui) {
+        ui.showAddBudget(amount, budgetList.getBudget());
+        budgetList.addToBudget(amount);
+        ui.showBudget(budgetList.getBudget());
```
Comment (RubaP) [inline src/main/java/duke/command/AddBudgetCommand.java]: Good example of following SLAP

File: src/main/java/duke/command/HelpCommand.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    /**
+     * Executes a command that adds the task into task list and outputs the result (GUI).
+     *
+     * @param items The task list that contains a list of tasks.
+     * @param ui To tell the user that it is added successfully.
+     * @return String to be outputted to the user.
+     */
+    @Override
+    public String executeGui(TaskList items, Ui ui) {
+        String str = Ui.helpRequest(items); //replace with events later on
```
Comment (RubaP) [inline src/main/java/duke/command/HelpCommand.java]: Comments are for the reader to explain the code, not for the developer to remind the leftover jobs

File: src/main/java/duke/command/UpdateCommand.java [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+    public void execute(TaskList items, Ui ui) {
+        try {
+            if (typeOfUpdate == 1) {
+                items.get(index).setDescription(taskDesc);
+            } else if (typeOfUpdate == 2) {
+                items.get(index).setDateTime(dateDesc);
+            } else if (typeOfUpdate == 3) {
+                Task newtaskObj = null;
+                if (typeDesc.equals("todo")) {
+                    if (items.get(index) instanceof Todo) {
```
Comment (RubaP) [inline src/main/java/duke/command/UpdateCommand.java]: Three levels of 'if' statements are here. Try to simplify this method since it is very hard to follow now

File: src/main/java/duke/command/UpdateCommand.java [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+                        throw new DukeException("     You are updating the same type of task! (Todo)");
+                    } else {
+                        newtaskObj = new Todo(items.get(index).getDescription());
+                    }
+                } else if (typeDesc.equals("deadline")) {
```
Comment (RubaP) [inline src/main/java/duke/command/UpdateCommand.java]: Third level of 'if' can be changed to switch block to make it more clear and extracting out any reusable code snippet as method.

File: src/main/java/duke/command/UpdateCommand.java [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+                Task newtaskObj = null;
+                if (typeDesc.equals("todo")) {
+                    if (items.get(index) instanceof Todo) {
+                        //throw new DukeException("     You are updating the same type of task! (Todo)");
+                        return "     (>_<) OOPS!!! You are updating the same type of task! (Todo)";
+                    } else {
+                        newtaskObj = new Todo(items.get(index).getDescription());
+                    }
+                } else if (typeDesc.equals("deadline")) {
+                    if (items.get(index) instanceof Event || items.get(index) instanceof Repeat) {
```
Comment (RubaP) [inline src/main/java/duke/command/UpdateCommand.java]: Same readability issue, if possible try to handle it in a more readable way (e.g. switch block, extract reusable code snippet as method)

File: src/main/java/duke/parser/DateParser.java [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+        String dayInString = startDate[0].split("/")[0];
+        String mthInString = startDate[0].split("/")[1];
+        String yrsInString = startDate[0].split("/")[2];
+
+        LocalDate localDate = LocalDate.of(Integer.parseInt(yrsInString),
+                                            Integer.parseInt(mthInString),
+                                            Integer.parseInt(dayInString));
+        switch (datePeriod) {
+        case "days" : localDate = localDate.plusDays(1);
```
Comment (RubaP) [inline src/main/java/duke/parser/DateParser.java]: Please define all the magical strings as constants

File: src/main/java/duke/storage/PriorityStorage.java [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+import java.io.IOException;
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.File;
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.util.ArrayList;
+
+public class PriorityStorage {
+    //protected String filePath = "./";
```
Comment (RubaP) [inline src/main/java/duke/storage/PriorityStorage.java]: Remove commented code

File: src/main/java/duke/storage/Storage.java [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+import java.util.ArrayList;
+
+/**
+ * Represents a storage to store the task list into a text file.
+ */
+public class Storage {
+    //protected String filePath = "./";
+    protected String filePath = "";   //27-28, 40-47
+    String storageClassPath = Storage.class.getProtectionDomain().getCodeSource().getLocation().getPath();
+    private static final int ZERO = 0;
```
Comment (RubaP) [inline src/main/java/duke/storage/Storage.java]: If these constants are repeatedly used by many classes, you can define it as an enum to increase the readability and maintainability

File: src/main/java/duke/storage/Storage.java [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+                if (commandList[ZERO].equals("T")) {
+                    if (taskDesc.trim().isEmpty()) {
+                        throw new DukeException("Error reading description, skipping to next line");
+                    } else {
+                        t = new Todo(taskDesc);
+                        t.setStatusIcon(checked);
+                        items.add(t);
+                    }
+                } else if (commandList[ZERO].equals("D")) {
+                    if (taskDesc.trim().isEmpty() || dateDesc.trim().isEmpty()) {
```
Comment (RubaP) [inline src/main/java/duke/storage/Storage.java:100]: So much of repeated code. Think of an effective way of reducing it by moving the repeated validation check to single place

File: src/main/java/duke/task/ContactList.java [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+     * Creates an updated contact list with the specified array list.
+     *
+     * @param contactList The updated array list.
+     */
+    public ContactList(ArrayList<Contacts> contactList) {
+        this.contactList = contactList;
+    }
+
+    /**
+     * To add a contact into the contact list.
```
Comment (RubaP) [inline src/main/java/duke/task/ContactList.java]: The first name of the method should be in simple present e.g. Adds

**Review by RubaP — COMMENTED** [review]
Overall, it is a good job. I havent repeatedly pointed out the same mistake. So everyone, please go through all the comments and address it if it is applicable to the code you wrote.
